lib/descendDisplay.ml

```diff
@@ -1282,6 +1288,580 @@ let print_aboville conf base max_level p =
+   L'algo fait un parcours PRE-ORDRE de l'arbre en donnant une colonne
+   initiale "logique" pour chaque personne, par exemple:  X-1 et X+1
+   pour les 2 enfants d'un pere (P) dans la colonne X.
+   Ce "X" initial peut etre augmenté (déplacé à droite) pour 2 raisons:
+
+   a) La case proposée est déjà occupée.  Un tableau global,
+   last_x, garde la position des derniers noeuds dans chaque rangée.
+   On modifie X = max( X, last_x[i]+2)  gardant au moins une colonne
+   vide entre chaque noeud.
+   b) Pour la meme raison, il se peut que les enfants de la personne
+   soient déplacés. ON termine en mettant X a mi-chemin entre les position
+   finale du premier et dernier enfant.
+
+    ***************************************************************** *)
+
+(* one td is [nb_cols * align * content]  see dag2Html.ml *)
+
+(* empty fill over n columns *)
+let td_fill x1 xn = [ (xn - x1, CenterA, TDnothing) ]
+
+(* hbar over xn - x1 columns. First and last are left/right and 50% *)
+let td_hbar x1 xn =
+  match xn - x1 with
+  | 0 -> [ (1, CenterA, TDnothing) ]
+  | 1 -> [ (1, CenterA, TDhr CenterA) ]
+  | _ ->
+      [ (1, LeftA, TDhr LeftA) ]
+      @ [ (xn - x1 - 1, CenterA, TDhr CenterA) ]
+      @ [ (1, RightA, TDhr RightA) ]
+
+(* regular cell, centered, with text as content (may contain |<br>) *)
+let td_cell cols align ip text flags =
+  match align with
+  | "center" -> [ (cols, CenterA, TDitem (ip, Adef.safe text, flags)) ]
+  | "right" -> [ (cols, RightA, TDitem (ip, Adef.safe text, flags)) ]
+  | "left" -> [ (cols, LeftA, TDitem (ip, Adef.safe text, flags)) ]
+  | _ -> [ (0, CenterA, TDitem (ip, Adef.safe text, flags)) ]
+
+(* tdal is   (int   *    list      ) list           *)
+(*           (lastx      list of td) list of rows   *)
+
+(* add elem to row ir. Update lastx *)
+let tdal_add tdal ir elem nx =
+  let tdal =
+    let rec loop tdal new_tdal ir elem nx =
+      match tdal with
+      | [] -> new_tdal
+      | (x, row) :: tdal ->
+          if ir = 0 then
+            let new_row = (nx, row @ elem) in
+            new_row :: loop tdal new_tdal (ir - 1) elem nx
+          else (x, row) :: loop tdal new_tdal (ir - 1) elem nx
+    in
+    loop tdal [] ir elem nx
+  in
+  tdal
+
+let get_bd_td_prop conf =
+  let bd = match Util.p_getint conf.env "bd" with Some x -> x | None -> 0 in
+  let td_prop =
+    match Util.p_getenv conf.env "td" with
+    | Some x -> " " ^ x
+    | _ -> (
+        match Util.p_getenv conf.env "color" with
+        | None | Some "" -> ""
+        | Some x -> " class=\"" ^ x ^ "\"")
+  in
+  (bd, td_prop)
+
+let reference conf base p s =
+  let iper = get_iper p in
+  if is_hidden p then s
+  else
+    String.concat ""
+      [
+        "<a href=\"";
+        (commd conf :> string);
+        (acces conf base p :> string);
+        "\" id=\"i";
+        string_of_iper iper;
+        "\"";
+        " class=\"normal_anchor\"";
+        " title=\"[go on individual page]\">";
+        s;
+        "</a>";
+      ]
```

Comment:
can you use `Printf.sprintf` instead?
It is more readable and It should be more efficient too

Reply:
Done